### ISSUE TITLE
use tput or similar

### ISSUE THREAD

**Issue by antonio-gg-dev** [TypedDevs/bashunit#247]
TODO #245

**Comment by h0adp0re** [comment]
As the pioneer of this topic, I can add that after implementing `tput` and patching some codes into `assert_equals_ignore_colors`, really only the snapshots failed across different terminal emulators, since they capture everything, including the environment-specific stuff.

---

In addition to already mentioned benefits, a specific scenario improved via the use of `tput` that is probably not too specific to me — committing from inside of neovim (via `fugitive.vim`, for example).
This is the pre-commit hook output in the message area with the hardcoded colors:

```plain
lib/bashunit
^[[1m^[[32mbashunit^[[0m - 0.10.1
Running scripts/__tests__/example_test.sh
^[[33m↷ Skipped^[[0m: Skipped test example
^[[2m    Not needed in this env^[[0m
^[[36m✒ Incomplete^[[0m: Incomplete test example
^[[2m    Add unit tests for scripts^[[0m
^[[32m✓ Passed^[[0m: Passing test example

^[[2mTests:     ^[[0m ^[[32m1 passed^[[0m, ^[[33m1 skipped^[[0m, ^[[36m1 incomplete^[[0m, 3 total
^[[2mAssertions:^[[0m ^[[32m1 passed^[[0m, ^[[33m1 skipped^[[0m, ^[[36m1 incomplete^[[0m, 3 total
^[[46mSome tests incomplete^[[0m
```

This is the output with `tput` implemented:

```plain
lib/bashunit
bashunit - 0.10.1
Running scripts/__tests__/example_test.sh
↷ Skipped: Skipped test example
    Not needed in this env
✓ Passed: Passing test example
✒ Incomplete: Incomplete test example
    Add unit tests for scripts

Tests:      1 passed, 1 skipped, 1 incomplete, 3 total
Assertions: 1 passed, 1 skipped, 1 incomplete, 3 total

 Some tests incomplete
```

This behavior is largely explained in the first post under "Portability" but to add more resolution — in a limited environment where terminal output is not expected or needed, `$TERM` is (hopefully) set to `dumb`, which means no fancy output please. This ensures exactly what my example highlights: _just in case_ there happens to be output, make it still readable by not printing any color codes.

**Comment by h0adp0re** [comment]
> I also have an observation to make after trying out tput, the GitHub actions went from being colored to black and white.

Yes, that is the effect of `TERM` being unset in that environment, which `tput` relies on to output _anything_.

There is surely a middle ground but it has not revealed itself yet.

**Comment by h0adp0re** [comment]
Ok, I've figured out a way to serve my usecase and leave the rest intact.

The solution is to slightly modify the function I added here so whenever `TERM` is explicitly `dumb`, it outputs nothing:

https://github.com/TypedDevs/bashunit/blob/46138132838dc7544296bd2760f1f461fa7b4eb6/src/colors.sh#L9-L18

This should keep colors in every other scenario, even when `TERM` is undefined:

```diff bash
  sgr() {
-   local codes=${1:-0}
+   local codes return
+ 
+   codes=${1:-0}
    shift
  
    for c in "$@"; do
      codes="$codes;$c"
    done
  
-   echo $'\e'"[${codes}m"
+   return=$'\e'"[${codes}m"
+ 
+   if [[ $TERM == 'dumb' ]]; then
+     return=""
+   fi
+ 
+   echo "$return"
  }
```

Simple as that, let me know if a PR like that would be welcome!

**Comment by Chemaclass** [comment]
Thanks, @h0adp0re ! Feel free to create a PR and we can continue the conversation from there 👍🏼

**Comment by Chemaclass** [comment]
@h0adp0re ping on this one? :)

**Comment by h0adp0re** [comment]
Sorry, I'm not sure if I'll be able to get to this anytime soon, as there are quite a few more urgent matters to attend to right now.

What happened to the couple of closed PRs?

**Comment by Chemaclass** [comment]
@h0adp0re I wasn't able to make it work https://github.com/TypedDevs/bashunit/pull/430#issuecomment-2955897567 , lot of failing tests for some reason that I didn't understand back then... maybe it's time to try this again nowadays :)
